### PR TITLE
Add fraud outcome elements to transaction details page

### PR DESCRIPTION
**Pull request by eduardoumpierre**
Fixes Automattic/woocommerce-payments-server#2793

#### Changes proposed in this Pull Request

Add the [new elements from the design](WIM5w2Y10SCNBFdMLSVftj-fi-1446-215272) to the transaction details page.

List of changes:
- New statuses
  - Payment blocked
  - Needs review
- New buttons
  - Block transaction
  - Approve transaction
- Add new entries to the timeline
  - Put on hold automatically
    - With the list of the rules that were triggered
  - Blocked automatically
    - With the list of the rules that were triggered
  - Manual block
  - Manual approval

#### Testing instructions

* **woocommerce-payments-server**
  * Checkout to https://github.com/Automattic/woocommerce-payments-server/pull/3153
  * Apply the following diff:
```diff
diff --git a/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php b/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php
index d352d46f0..1fe82807f 100644
--- a/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php
+++ b/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php
@@ -22,6 +22,7 @@ use WCPay\Service\Payment_Methods_Service;
 use WCPay\Service\Promotion_Service;
 use WCPay\Service\Sift_Service;
 use WCPay\Service\Payment_Authorization_Service;
+use WCPay\Service\Fraud_Outcome_Service;
 use WCPay\Stripe\Stripe_Client;
 use WCPay\Utils\Config_Interface;
 use WCPay\Utils\Currency;
@@ -91,6 +92,13 @@ class Intentions_Controller extends Base_Controller {
 	 */
 	private $payment_authorization_service;
 
+	/**
+	 * Fraud Outcome Service
+	 *
+	 * @var Fraud_Outcome_Service
+	 */
+	private $fraud_outcome_service;
+
 	/**
 	 * Route config.
 	 *
@@ -238,6 +246,7 @@ class Intentions_Controller extends Base_Controller {
 	 * @param WooPay_Service                $woopay_service WooPay Service.
 	 * @param Charge_Service                $charge_service Charge Service.
 	 * @param Payment_Authorization_Service $payment_authorization_service Payment Authorization Service.
+	 * @param Fraud_Outcome_Service         $fraud_outcome_service Fraud Outcome Service.
 	 */
 	public function __construct(
 		Config_Interface $config,
@@ -252,7 +261,8 @@ class Intentions_Controller extends Base_Controller {
 		Payment_Fingerprint_Service $payment_fingerprint_service,
 		WooPay_Service $woopay_service,
 		Charge_Service $charge_service,
-		Payment_Authorization_Service $payment_authorization_service
+		Payment_Authorization_Service $payment_authorization_service,
+		Fraud_Outcome_Service $fraud_outcome_service
 	) {
 		parent::__construct( $config, $tracker, $logger, $stripe_client, $account_service );
 		$this->fee_service                   = $fee_service;
@@ -263,6 +273,7 @@ class Intentions_Controller extends Base_Controller {
 		$this->woopay_service                = $woopay_service;
 		$this->charge_service                = $charge_service;
 		$this->payment_authorization_service = $payment_authorization_service;
+		$this->fraud_outcome_service         = $fraud_outcome_service;
 	}
 
 	/**
@@ -759,7 +770,7 @@ class Intentions_Controller extends Base_Controller {
 		}
 
 		// @todo Retrieve merchant ruleset setting.
-		$ruleset = [];
+		$ruleset = [ true ];
 
 		// We only want to apply the rules if the client version it at least 5.7.0 and
 		// the merchant ruleset setting is available.
@@ -779,6 +790,17 @@ class Intentions_Controller extends Base_Controller {
 		 */
 		$fraud_outcome = strtolower( explode( ' ', $request['metadata']['customer_name'] )[1] );
 
+		$account = $this->account_service->get_account_for_current_blog_or_error( $request->is_test() );
+		$outcome = [
+			'order_id'          => $request['metadata']['order_id'],
+			'payment_intent_id' => $request->get_param( 'intention_id' ) ?? '',
+			'ruleset_id'        => 1,
+			'ruleset_results'   => '[]',
+			'status'            => $fraud_outcome,
+		];
+
+		$this->fraud_outcome_service->save_fraud_outcome( $account, $outcome );
+
 		switch ( $fraud_outcome ) {
 			case Fraud_Outcome::FRAUD_OUTCOME_BLOCK:
 				$this->logger->log_debug( 'Risk rules: Payment intent blocked', [ 'request' => (array) $request ] );

```

* **woocommerce-payments**
  * Checkout to this branch  `add/2793-transaction-details`
  * Build your local version `npm run start`
  * Listen to Stripe webhooks
  * Apply the following diff:
```diff
diff --git a/includes/class-wc-payments.php b/includes/class-wc-payments.php
index 00bab27c..d7426bed 100644
--- a/includes/class-wc-payments.php
+++ b/includes/class-wc-payments.php
@@ -566,6 +566,19 @@ class WC_Payments {
 
 		add_action( 'admin_enqueue_scripts', [ __CLASS__, 'enqueue_assets_script' ] );
 		add_action( 'wp_enqueue_scripts', [ __CLASS__, 'enqueue_assets_script' ] );
+
+		add_filter(
+			'wcpay_api_request_params',
+			function( $params, $api, $method ) {
+				if ( false !== strpos( $api, WC_Payments_API_Client::INTENTIONS_API ) && WC_Payments_API_Client::POST === $method ) {
+					$params['fraud_settings_enabled'] = 'true';
+				}
+
+				return $params;
+			},
+			10,
+			3
+		);
 	}
 
 	/**
diff --git a/woocommerce-payments.php b/woocommerce-payments.php
index 69de822b..b1dd530a 100644
--- a/woocommerce-payments.php
+++ b/woocommerce-payments.php
@@ -12,7 +12,7 @@
  * WC tested up to: 7.3.0
  * Requires at least: 5.9
  * Requires PHP: 7.0
- * Version: 5.6.2
+ * Version: 5.7.0
  *
  * @package WooCommerce\Payments
  */

```

* Perform 2 on-hold orders
    * Open your test shop
    * Add an item to the shopping cart, go to the checkout page
    * The "Last name" should be Review
    * Use the card 4242 4242 4242 4242
* Go to **WooCommerce > Orders** and click on one of the new orders
* Click on the payment_intent link to open the transaction details page
* You should see something similar to the following image
image

* Please, confirm the following elements:
  * **Header**
    * There should be two buttons "Block transaction" and "Approve transaction"
    * The status should be "Needs review"
    * The capture notice shouldn't contain a "Capture button" and the text should include "Approving this transaction will capture the charge."
  * **Timeline**
    * There should be a "**Payment was screened by your fraud filters and placed in review.**" entry
  * Click on "Block transaction"  
noname

* Please, confirm the following elements:
  * **Header**
    * The "Block transaction" button is in a loading state
    * The "Approve transaction" button is disabled
* A notice will show up and the UI will be updated

image

* Please, confirm the following elements:
  * **Header**
    * The buttons aren't visible anymore
    * The status is now "Payment blocked"
    * The capture notice is not visible anymore
  * **Timeline**
    * The timeline includes "Payment was blocked by your_user"

* Go back to the orders list, open the second order you've created, and open the transaction details page
* Check the same elements for the header and timeline
* Click on the "Approve transaction" button
* It should be in a loading state and the "Block transaction" button should be disabled
* A notice should pop up with a success message
image

* Please, confirm the following elements:
  * **Header**
    * The buttons aren't visible anymore
    * The status is now "Paid"
    * The capture notice is not visible anymore
  * **Timeline**
    * The timeline includes "Payment was approved by your_user"

### Timeline

* Checkout to [Taha's PR](https://github.com/Automattic/woocommerce-payments/pull/5837) – both client and server
* Adjust the protection rules to review
* Perform an order that should be ended as review
* Checkout back to this branch – both client and server
* Open the transaction details of the order you've just created
* It should contain the list of the rules that were triggered
image

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5841` or branch name `add/2793-transaction-details` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/2793-transaction-details) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: d9618c26e47ff34f336adb5fd4e8a2103a48e65e
- Build time: 2023-03-24 17:50:05 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/data/fraud-outcomes/test/resolvers.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { apiFetch, dispatch } from '@wordpress/data-controls';
```
Comment (tpaksu) [inline client/data/fraud-outcomes/test/resolvers.ts]: GH Actions thinks it's unused.

**Review by tpaksu — APPROVED** [review]
Approving this because the testing instructions work as expected, the code looks good, there's one thing that GH Actions complains, and there's the fraud outcome shown again after capturing the payment, but you created another issue for that. LGTM! Thanks for working on this with such detail!

File: client/data/fraud-outcomes/test/resolvers.ts (hunk 1 above)
Comment (eduardoumpierre) [reply client/data/fraud-outcomes/test/resolvers.ts]: Removed on [16798c2](https://github.com/Automattic/woocommerce-payments/pull/5841/commits/16798c2014b6b49a82d9d4bdfabd52ec9793072e).

**Pull request merged**